archaius2-core/src/main/java/com/netflix/archaius/config/AbstractConfig.java

```diff
@@ -225,243 +227,158 @@ public Config getPrefixedView(String prefix) {
         return result;
     }
 
-    @Override
-    public Long getLong(String key) {
-        String value = getString(key);
-        if (value == null) 
+    protected <T> T getValue(Class<T> type, String key) {
+        Object rawProp = getRawProperty(key);
+        if (rawProp == null) {
             return notFound(key);
-        try {
-            return decoder.decode(Long.class, value);
         }
-        catch (NumberFormatException e) {
-            return parseError(key, value, e);
+        if (rawProp instanceof String) {
+            try {
+                String value = interpolator.create(lookup).resolve(rawProp.toString());
+                if (type.equals(boolean.class) || type.equals(Boolean.class)) {
+                    if (value.equalsIgnoreCase("true") || value.equalsIgnoreCase("yes") || value.equalsIgnoreCase("on")) {
+                        return (T) Boolean.TRUE;
+                    }
+                    else if (value.equalsIgnoreCase("false") || value.equalsIgnoreCase("no") || value.equalsIgnoreCase("off")) {
+                        return (T) Boolean.FALSE;
+                    }
+                    return parseError(key, value, new Exception("Expected one of [true, yes, on, false, no, off]"));
+                }
+                return decoder.decode(type, value);
+            } catch (NumberFormatException e) {
+                return parseError(key, rawProp.toString(), e);
+            }
+        } else if (type.isInstance(rawProp)) {
+            return (T)rawProp;
+        } else {
+            return parseError(key, rawProp.toString(),
+                    new NumberFormatException("Property " + rawProp.toString() + " is of wrong format " + type.getCanonicalName()));
         }
     }
 
     @Override
+    public Long getLong(String key) {
+        return getValue(Long.class, key);
+    }
+
+    @Override
     public Long getLong(String key, Long defaultValue) {
-        String value = getString(key, null);
-        if (value == null) 
-            return defaultValue;
         try {
-            return decoder.decode(Long.class, value);
-        }
-        catch (NumberFormatException e) {
-            return parseError(key, value, e);
+            return getValue(Long.class, key);
```

Comment:
Can we create a variant of getValue(Class<T> type, String key, T defaultValue) that will handle the NoSuchElementException instead of having to handle the exception in all the getXXX methods?